app/components/data-download.tsx

```diff
@@ -98,43 +62,162 @@ const DataDownloadInner = memo(
+    chartConfig,
+  }: {
+    dataSetIri: string;
+    title: string;
+    chartConfig: ChartConfig;
+  }) => {
+    const allPrepared = usePreparedData({
+      dataSetIri,
+      chartConfig,
+      extent: "all",
+    });
+    const visiblePrepared = usePreparedData({
+      dataSetIri,
+      chartConfig,
+      extent: "visible",
+    });
+    const popupState = usePopupState({
+      variant: "popover",
+      popupId: "dataDownloadMenu",
+    });
 
     return (
-      <DownloadButton onClick={() => downloadData()}>
-        {extent === "visible" ? (
-          <Trans id="button.download.data.visible">Download visible data</Trans>
-        ) : (
-          <Trans id="button.download.data.all">Download all data</Trans>
-        )}
-      </DownloadButton>
+      <>
+        <DownloadButton iconName="download" {...bindHover(popupState)}>
+          <Trans id="button.download">Download</Trans>
+        </DownloadButton>
+        <HoverMenu
+          {...bindMenu(popupState)}
+          anchorOrigin={{ vertical: "top", horizontal: "center" }}
+          transformOrigin={{ vertical: "bottom", horizontal: "center" }}
+        >
+          <Typography
+            sx={{
+              ...PADDING_PROPS,
+              borderBottom: "1px solid",
+              borderBottomColor: "grey.500",
+            }}
+          >
+            Download
+          </Typography>
+          {OPTIONS_TO_RENDER.map((d) => (
+            <DownloadMenuItem
+              key={d.extent + d.fileFormat}
+              title={title}
+              extent={d.extent}
+              fileFormat={d.fileFormat}
+              closeMenu={popupState.close}
+              {...(d.extent === "all" ? allPrepared : visiblePrepared)}
+            />
+          ))}
+        </HoverMenu>
+      </>
     );
   }
 );
 
-const RunSparqlQuery = ({
-  url,
+const DownloadMenuItem = ({
+  title,
+  data,
+  columnKeys,
   extent,
+  fileFormat,
+  closeMenu,
 }: {
-  url: string;
-  extent: DataDownloadExtent;
+  title: string;
+  data: Observation[];
+  columnKeys: string[];
+  extent: Extent;
+  fileFormat: FileFormat;
+  closeMenu: () => void;
+}) => {
+  const download = useCallback(() => {
+    return fetch("/api/download", {
+      method: "POST",
+      headers: {
+        "Content-Type": "application/json",
+      },
+      body: JSON.stringify({
+        columnKeys,
+        data,
+        fileFormat,
+      }),
+    }).then((res) =>
+      res.blob().then((blob) => saveAs(blob, `${title}.${fileFormat}`))
+    );
+  }, [columnKeys, data, fileFormat, title]);
+
+  return (
+    <DownloadMenuItemInner onClick={() => download().then(() => closeMenu())}>
+      {extent === "visible" ? (
+        <>
+          <Trans id="button.download.data.visible">Overlooking dataset</Trans> (
+          {fileFormat.toUpperCase()})
+        </>
+      ) : (
+        <>
+          <Trans id="button.download.data.all">Full dataset</Trans> (
+          {fileFormat.toUpperCase()})
+        </>
```

Comment:
While download is happening, it seems the user has no visual feedback ? Maybe a spinner could be added while download() runs ?

Reply:
Yes, this is on my TODO list 😅